test/micro/org/openjdk/bench/java/math/BigIntegerSquareRoot.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) 2014, 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package org.openjdk.bench.java.math;
+
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.BenchmarkMode;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.OperationsPerInvocation;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Warmup;
+import org.openjdk.jmh.infra.Blackhole;
+
+import java.math.BigInteger;
+import java.util.Random;
+import java.util.concurrent.TimeUnit;
+
+@BenchmarkMode(Mode.AverageTime)
+@OutputTimeUnit(TimeUnit.NANOSECONDS)
+@State(Scope.Thread)
+@Warmup(iterations = 5, time = 1)
+@Measurement(iterations = 5, time = 1)
+@Fork(value = 3)
+public class BigIntegerSquareRoot {
+
+    private BigInteger[] hugeArray, largeArray, smallArray;
+    private static final int TESTSIZE = 1000;
+
+    @Setup
+    public void setup() {
+        Random r = new Random(1123);
+        int numbits = r.nextInt(16384);
+
+        hugeArray = new BigInteger[TESTSIZE]; /*
+         * Huge numbers larger than
+         * MAX_LONG
+         */
+        largeArray = new BigInteger[TESTSIZE]; /*
+         * Large numbers less than
+         * MAX_LONG but larger than
+         * MAX_INT
+         */
+        smallArray = new BigInteger[TESTSIZE]; /*
+         * Small number less than
+         * MAX_INT
+         */
+
+        for (int i = 0; i < TESTSIZE; i++) {
+            int value = Math.abs(r.nextInt());
```

Comment:
There's a risk of an overflow here if the random `int` is `MIN_VALUE`, which would throw an exception later.

Reply:
It wasn't the overflow by itself that worried me, but that a later invocation of `sqrt*()` could throw.

Again, the "huge" numbers are less than 2^66. You might to take a look at `shiftArray` in `BigInteger.java` for inspiration, and adding some leading `0` bits to exercise the normalization/denormalization impact.

Reply:
Actually, the problem rises already when constructing the BigInteger, by invalid decimal string construction.